docs/refactor_notes.md

```diff
@@ -0,0 +1,139 @@
+# refactor notes
+
+## FLow Layer
+
+The flow layer needs to be refactored in full, it has to be seperated into smaller more defined components.
+
+This refactor has the following "phases"
+
+1. review relations
+---
+
+* list the current components inside the flow layer
+* list the responsibilities of each component
+* list the relations between each component
+
+---
+2. modify relations
+---
+
+* depulicate aggregate responsibilities over all components
+* re organize responsibilities into distinct "osi stack" like layers
+* re structure components into layers such that they only talk to layers above and below
+
+---
+3. spec out interfaces
+---
+
+* given the new component relations define new compilation targets (liblokinet-whatever) and set up the high level deps
+* in each component define interface types for each responsibility with their relations outside the component itself
+* define internal component relations
+* define unit tests for each responsibility in each component
+
+---
+4. refactor
+---
+
+* rearrange the code from the existing pile into the new structure
+* add unit tests before moving code into place to ensure it does what i think it does
+
+
+## Concrete Plan
+
+---
+1. define new interfaces and structure
+---
+
+* define new cmake libraries that split up lokinet-amalgum more
+* define new interface types in headers in each new split up library
+
+---
+2. wire up old implementation to new interface
+---
+
+* add unit tests that can fail when wired up to old code
+* wire up old code to new interface
+* verify working state
+
+---
+3. add new implementation to new interface
+---
+
+* implement new interface so that unit tests pass with new implementation
+* verify working state
+
+
+
+### new interface types (1)
+
+the component to refactor is `llarp::service::Context`
+
+the types that will get a new interface type are:
+
+`llarp::service::Context` -> `llarp::service::I_ServiceContext`
+
+`llarp::service::Endpoint` -> `llarp::service::I_Endpoint`
+
+`llarp::handlers` namespace will be deprecated and no longer publicly exposed
+
+an amalgum interface for obtaining all platform specific details, like event loop, vpn platform, etc: `llarp::service::I_PlatformDetails`
```

Comment:
FWIW I'm not a fan of the "I" prefix, partly because there really is no such thing as a pure interface in C++ (i.e. it's purely convention).  Interface classes, though, are awkwardly restricted in many classes.  For instance, I could easily imagine this `I_PlatformDetails` having some protected member variables/functions that de-duplicate common functionality from platform code, but as soon as you have any member or non-pure-virtual base implementation it isn't really an interface anymore.

Reply:
so a `WhateverThing_Base` as an abstract class would be better then?